modules/memfault-firmware-sdk/memfault_gnss_metrics.c

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <modem/location.h>
+#include "memfault/metrics/metrics.h"
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(memfault_ncs_metrics, CONFIG_MEMFAULT_NCS_LOG_LEVEL);
+
+static bool session_in_progress = false;
+static location_event_handler_t real_location_event_handler;
+
+static void memfault_location_event_handler(const struct location_event_data *event_data);
+static void memfault_gnss_metrics_start_fix_session(void);
+static void memfault_gnss_metrics_stop_fix_session(const struct location_event_data *event_data);
+
+// To maintain separation of metrics collection from NCS library code, we need a way to
+// know when a fix request is made, and when the result of the request is ready (fix or error).
```

Comment:
Why do you need to know when the fix request is made? Is it just to be able to start a timer to measure the fix time?

Reply:
We are currently using the session's elapsed time to determine the fix time, but needing to know when the fix request is made goes beyond the collection of the fix time metric. We need that information in order to start a fix "session".

We believe location-related metrics best live under the umbrella of what we call a "session". This is a pretty new feature we have (we are still working on public docs for it!), and so I'll give some background in case you are unfamiliar. We are excited to introduce this to NCS. Some metrics, like a location fix time, only make sense when captured within a specific time window, such as the window that starts with a fix request and ends when either a fix or a timeout. Collecting metrics outside a session would mean they are grouped into a periodic "heartbeat" that typically spans multiple hours, and therefore could span multiple sessions. Insights that could be really interesting like the average fix time could not be computed if the data were reported via a heartbeat since we would lose data from individual sessions. (Note that the `debug_module` in the asset tracker app tries to work around this by [triggering a heartbeat manually after location data is collected](https://github.com/NordicSemiconductor/asset-tracker-cloud-firmware-aws/blob/971ba026a37f5ad7df82f6d39dcd36aef0e80e70/src/modules/debug_module.c#L293), but this is not the intended use of heartbeats. We need a better way to collect location data that doesn't use heartbeats.)

To use sessions, we need to start and stop the session each with one API call. Any metrics collected in the window will then be grouped together and reported as that session's data. For a location acquisition session, therefore, we need to know when the application begins to look for a fix, and when it stops looking for a fix.  In the location library, it looks like `location_request()` function is the way the application requests new location data. Some ideas I have for how we can adjust the location library to find out when to start a fix session:
- Create a new event type under the `location_event_id` enumeration called something like `LOCATION_EVT_SEARCH_INITIATED`, then call the event handler within `location_core_location_get_pos()`. That should catch both periodic and one-off location requests.
- Allow a different callback to be registered with the location library that will similarly be called whenever the location core initiates a location get. To avoid changing the location library's API, this handler could be passed in via the `location_config` structure.

What do you think?

Reply:
What if the location event would tell the fix time and other details that you'd like to collect? This way you wouldn't need to care about when the `location_request()` was called.

Calculating the fix time from the `location_request()` call to the `LOCATION_EVT_LOCATION` is not necessarily the best measure to evaluate GNSS performance because the GNSS module is not active all this time. There are certain conditions when GNSS is started and this could be taken into account in the Location library when it reports fix time.

Reply:
@trantanen Unfortunately without knowing when the request starts, we won't be able to encapsulate these metrics in a session. I'm seeing that building out of the box location metrics is going to be more complicated than I originally scoped, and instead we may require the user to call the session start themselves. Let me consider some alternatives and I will get back to you on this!

Reply:
> Calculating the fix time from the location_request() call to the LOCATION_EVT_LOCATION is not necessarily the best measure to evaluate GNSS performance because the GNSS module is not active all this time. 

Ah thank you for this correction, I see now that the way I'm calculating gnss fix time may not equate to gnss active time.

> There are certain conditions when GNSS is started and this could be taken into account in the Location library when it reports fix time.

I see -- so we could have a fix time associated with the gnss that maps to how long the GNSS was active for, as well as an overall fix time, which may include the use of other methods?

Reply:
I was reflecting on how we are able to calculate LTE time to connect and it is because we register a handler with the `lte_lc_trace` module which is then used in the `lte_lc` library when certain events occur like putting the modem in a certain mode. Would you consider having a similar tracing mechanism for the location library (`location_trace.h`/`location_trace.c`), which would call the trace callback when various events of interest occur?  These traces could be captured at the `location_core` level when issuing a location get and when a timeout occurs (for starters). The associated method involved would then be passed to the callback. For more granular information, the tracing could also occur at the method level, giving deeper insight into communications with the hardware. What do you think?